src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,580 @@
+        out.close();
+        return content;
+    }
+    
+    private Map<String, String> readHeaders(MessageUnpacker unpacker) throws IOException {
+        int headerCount = unpacker.unpackMapHeader();
+        if (headerCount > 0) {
+            Map<String, String> headers = new HashMap<String, String>();
+            for (int i = 0; i < headerCount; i++) {
+                headers.put(unpacker.unpackString(), unpacker.unpackString());
+            }
+            return headers;
+        } else {
+            return null;
+        }
+    }
+
+    private void writeHeaders(Map<String, String> headers, MessagePacker packer) throws IOException {
+        if (headers != null) {
+            packer.packMapHeader(headers.size());
+            for (String k: headers.keySet()) {
+                packer.packString(k);
+                packer.packString(headers.get(k));
+            }    
+        } else {
+            packer.packMapHeader(0);
+        }
+    }
+    
+    private Collection<String> readStreamIds(MessageUnpacker unpacker) throws IOException {
+        int streamCount = unpacker.unpackArrayHeader();
+        Collection<String> streams = null;
+        
+        if (streamCount > 0) {
+            streams = new ArrayList<String>();
+            for (int i = 0; i < streamCount; i++) {
+                streams.add(unpacker.unpackString());
+            }
+        }
+        
+        return streams;
+    }
+    
+    private void writeStreamIds(Collection<String> streamIds, MessagePacker packer) throws IOException {
+        if (streamIds != null) {
+            packer.packArrayHeader(streamIds.size());
+            for (String s: streamIds) {
+                packer.packString(s);
+            }
+        } else {
+            packer.packArrayHeader(0);
+        }
+    }
+    
+    private Object[] bindArguments(MessageUnpacker unpacker, List<Class<?>> paramTypes) throws IOException {
+        int argumentCount = unpacker.unpackArrayHeader();
+        
+        if (paramTypes.size() != argumentCount) {
+            throw new RuntimeException(String.format("Invocation provides %d argument(s) but target expects %d.", argumentCount, paramTypes.size()));
+        }
+        
+        Object[] arguments = new Object[argumentCount];
+        
+        for (int i = 0; i < argumentCount; i++) {
+            arguments[i] = readValue(unpacker, paramTypes.get(i));
+        }
+        
+        return arguments;
+    }
+    
+    private Object readValue(MessageUnpacker unpacker, Class<?> itemType) throws IOException {
+        // This shouldn't ever get called with itemType == null, but we return anyways to avoid NullPointerExceptions
+        if (itemType == null) {
+            return null;
+        }
+        MessageFormat messageFormat = unpacker.getNextFormat();
+        ValueType valueType = messageFormat.getValueType();
+        int length;
+        Object item = null;
+         
+        switch(valueType) {
```

Comment:
I don't account for the MessagePack data being malformed - if the user passes a Map when the function expects a String, this should just throw a MessagePackException. The C# client appears to do the same

Reply:
How does this surface to the HubConnection user?

Reply:
A call to `ParseMessages` would throw a RuntimeException w/ "Error reading messagepack data" - JSONHubProtocol does the same: 

https://github.com/dotnet/aspnetcore/blob/0d156487aa27df65d68fc7fa31e7b6cb9f4d7581/src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/JsonHubProtocol.java#L175-L177

While the C# Client's method is `TryParseMessages`, so it just returns false if it runs into issues.